x/go/calc/calc.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2023 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+package calc
+
+import (
+	"errors"
+	"go/ast"
+	"go/token"
+	"math"
+	"strconv"
+
+	"github.com/synnaxlabs/x/stack"
+)
+
+type Expression struct {
+	exp ast.Expr
+}
+
+type Resolver interface {
+	Resolve(string) (float64, error)
+}
+
+func findTokens(s string) ([]string, error) {
+	tokens := []string{}
+	currToken := ""
+	for i, c := range s {
+		if c == ' ' {
+			if currToken != "" {
+				tokens = append(tokens, currToken)
+				currToken = ""
+			}
+			continue
+		}
+		//	Check if token is a negative number
+		if c == '-' && currToken == "" && (i == 0 || tokens[len(tokens)-1] == "(" || tokens[len(tokens)-1] == "+" || tokens[len(tokens)-1] == "-" || tokens[len(tokens)-1] == "*" || tokens[len(tokens)-1] == "/" || tokens[len(tokens)-1] == "^") {
```

Comment:
No need to repeatedly access `tokens[len(tokens-1)]`. Put it into a temporary variable instead. Also, instead of explicitly doing if checks on each valid token, you can create a private, global map and check if the token is in the map. Not only will this be cleaner, it'll also be faster.